src/app/autocomplete/autocomplete.component.ts

```diff
@@ -1,62 +1,56 @@
-import { WorkshopRepository } from '../services/workshops/workshopRepository'
+import { Component, OnInit, EventEmitter, Output } from '@angular/core';
+import { WorkshopRepository, ILocation } from '../services/workshops/workshopRepository'
 
 @Component({
     selector: 'autocomplete',
-    host: { '(document:click)': 'handleClick($event)' },
     templateUrl: './autocomplete.component.html',
     styleUrls: ['./autocomplete.component.scss']
 })
 export class AutocompleteComponent {
 
     public query = '';
-    public countries = [];
-    public filteredList = [];
-    public elementRef;
+    public countries: ILocation[];
     private workshopRepository: WorkshopRepository;
 
     @Output() selectionChanged = new EventEmitter();
 
-    constructor(myElement: ElementRef, workshopRepo: WorkshopRepository) {
-        this.elementRef = myElement;
+    constructor(private workshopRepo: WorkshopRepository) {
         this.workshopRepository = workshopRepo;
     }
 
     ngOnInit() {
-        this.workshopRepository.getLocations().then(loc => {
-            for (var i = 0; i < loc.length; i++) {
-                this.countries.push(loc[i].name);
-            }
+        this.countries = [];
+        this.workshopRepository.getLocations().then((loc: ILocation[]) => {
+            this.countries = loc;
         });
     }
 
-    filter() {
-        if (this.query !== "") {
-            this.filteredList = this.countries.filter(function (el) {
-                return el.toLowerCase().indexOf(this.query.toLowerCase()) > -1;
-            }.bind(this));
-        } else {
-            this.filteredList = [];
+    public select(value: any) {
+        if (value) {
+            let locationId = value.id;
+            if (Number.isNaN(locationId)) {
+                this.query = "";
+            }
+            else if (value.id) {
+                this.workshopRepository.getLocations().then((locations: ILocation[]) => {
+                    locations.forEach((loc: ILocation) => {
+                        if (locationId === loc.id) {
+                            this.query = loc.name;
+                        }
+                    })
+                });
+            }
         }
-    }
 
-    public select(item) {
-        this.query = item;
-        this.filteredList = [];
+        if (value && value.item) {
+            this.query = value.item.name;
+            this.selectionChanged.emit(value.item);
+        }
     }
 
-    handleClick(event) {
-        var clickedComponent = event.target;
-        var inside = false;
-        do {
-            if (clickedComponent === this.elementRef.nativeElement) {
-                inside = true;
-            }
-            clickedComponent = clickedComponent.parentNode;
-        } while (clickedComponent);
-        if (!inside) {
-            this.filteredList = [];
-            this.selectionChanged.emit(this.query);
+    public onBlur(value: string) {
+        if (value === "") {
```

Comment:
what if the earlier value was also empty? it'll cause an unnecessary call, right?